src/StreamUtils.bsv

```diff
@@ -0,0 +1,367 @@
+function Tuple3#(DataStream, DataStream, DataBytePtr) getConcatStream (DataStream streamA, DataStream streamB, DataBytePtr bytePtrA, DataBytePtr bytePtrB);
+    Bool isCallLegally = (streamA.isLast && bytePtrA <= getMaxBytePtr() && bytePtrA > 0);
+    DataBitPtr bitPtrA = zeroExtend(bytePtrA) << fromInteger(valueOf(BYTE_WIDTH_WIDTH));
+
+    // Fill the low PtrA bytes by streamA data
+    Data concatDataA = streamA.data;
+    ByteEn concatByteEnA = streamA.byteEn;
+
+    // Fill the high bytes by streamB data
+    Data concatDataB = streamB.data << bitPtrA;
+    ByteEn concatByteEnB = streamB.byteEn << bytePtrA;
+    Data concatData = concatDataA | concatDataB;
+    ByteEn concatByteEn = concatByteEnA | concatByteEnB;
+
+    // Get the remain bytes of streamB data
+    DataBitPtr resBitPtr = getMaxBitPtr() - bitPtrA;
+    DataBytePtr resBytePtr = getMaxBytePtr() - bytePtrA;
+    Data remainData = streamB.data >> resBitPtr;
+    ByteEn remainByteEn = streamB.byteEn >> resBytePtr;
+
+    // Get if the concat frame is the last
+    Bool isConcatStreamLast = streamB.isLast;
+    DataBytePtr remainBytePtr = 0;
+    if (resBytePtr < bytePtrB ) begin    
+        isConcatStreamLast = False;
+        remainBytePtr = bytePtrB - resBytePtr;
+    end
+     DataStream concatStream = getEmptyStream;
+     DataStream remainStream = getEmptyStream;
+
+    // package the return concatStream and remainStream
+    if(isCallLegally) begin
+        concatStream = DataStream{
+            data: concatData,
+            byteEn: concatByteEn,
+            isFirst: streamA.isFirst,
+            isLast: isConcatStreamLast
+        };
+        remainStream = DataStream{
+            data: remainData,
+            byteEn: remainByteEn,
+            isFirst: False,
+            isLast: True
+        };
+    end
+    return tuple3(concatStream, remainStream, remainBytePtr);
+endfunction
+
+function Action showDataStream (DataStream stream);
+    return action
+        $display("   Data = %h", stream.data);
+        $display("   byteEn = %b", stream.byteEn);
+        $display("   isFirst = %b, isLast = %b", stream.isFirst, stream.isLast);
+    endaction;
+endfunction
+
+function Action checkDataStream (DataStream stream, String name);
+    if (stream.byteEn == 0 || stream.data == 0) begin 
+        return action
+            $display("Error: empty dataStream ", name);
+            showDataStream(stream);
+            $finish();
+        endaction;
+    end 
+    else begin
+        return action
+        endaction;
+    end
+endfunction
+
+(* synthesize *)
+module mkStreamConcat (StreamConcat ifc);
+
+    FIFOF#(DataStream) inputFifoA <- mkFIFOF;
+    FIFOF#(DataStream) inputFifoB <- mkFIFOF;
+    FIFOF#(DataStream) outputFifo <- mkFIFOF;
+
+    FIFOF#(StreamWithPtr) prepareFifoA <- mkFIFOF;
+    FIFOF#(StreamWithPtr) prepareFifoB <- mkFIFOF;
```

Comment:
The above 5 lines of code looks good.

you may ask why we need to make code pretty. In fact, it's not a problem of pretty or not, but aligned code can make reviewer's life easy. When you will review a lot of code, you will know what it feels like. For better aligned code, you can glimpse the code and won't think much of it, if the code is badly aligned, you brain will take extra effort to "decode" the code you are reading, which will make the reviewer tired for a lot of code to review.

For example, we do not allow to write code inline, and force you to write each element in a new line in an enum or struct. This is because when you are reviewing, your eyes will first look top down, trying to find the boundary of a piece of code. For an `if` statement, if you write like:
```
if () begin
    // xxxxx
end
else begin
    // xxxxx
end
```
you will quickly get the boundary and with a simple glimpse you know the if statement has a else branch. But if you write
```
if () begin
    // xxxxx
end else begin
    // xxxxx
end
```
your eyes have to move right a little to find that there is a else statement. This is extra work for your brain.

Reply:
Understand. I will pay attention to the code format.